monai/engines/workflow.py

```diff
@@ -0,0 +1,121 @@
+            save checkpoint into files.
+        additional_metrics (list): more ignite metrics that also attach to Ignite Engine.
+        handlers (list): every handler is a set of Ignite Event-Handlers, like:
+            CheckpointHandler, StatsHandler, TimerHandler, etc.
+
+    """
+
+    def __init__(
+        self,
+        device,
+        max_epochs,
+        amp,
+        data_loader,
+        prepare_batch=default_prepare_batch,
+        key_metric=None,
+        additional_metrics=None,
+        handlers=None,
+    ):
+        # FIXME:
+        if amp:
+            print("Will add AMP support when PyTorch v1.6 released.")
+        assert isinstance(device, torch.device), "must provide PyTorch device information."
+        assert isinstance(max_epochs, int), "must set max epoch number."
+        assert isinstance(data_loader, torch.utils.data.DataLoader), "data_loader must be PyTorch DataLoader."
+
+        self.engine = Engine(self._iteration)
+        # set all sharable data for the workflow based on Ignite engine.state
+        self.engine.state = State(
+            seed=0,
+            iteration=0,
+            epoch=0,
+            max_epochs=max_epochs,
+            epoch_length=-1,
+            output=None,
+            batch=None,
+            metrics={},
+            dataloader=None,
+            device=device,
+            amp=amp,
+            key_metric_name=None,  # we can set many metrics, only use key_metric to compare and save the best model
+            best_metric=-1,
+            best_metric_epoch=-1,
+        )
+        self.data_loader = data_loader
+        self.prepare_batch = prepare_batch
+
+        metrics = None
+        if key_metric is not None:
+            assert isinstance(key_metric, dict), "key_metric must be a dict object."
+            self.engine.state.key_metric_name = list(key_metric.keys())[0]
+            metrics = key_metric
+            if additional_metrics is not None and len(additional_metrics) > 0:
+                assert isinstance(additional_metrics, dict), "additional_metrics must be a dict object."
+                metrics.update(additional_metrics)
+            for name, metric in metrics.items():
+                metric.attach(self.engine, name)
+
+            @self.engine.on(Events.EPOCH_COMPLETED)
+            def post_epoch_process(engine):
+                if engine.state.key_metric_name is not None:
+                    current_val_metric = engine.state.metrics[engine.state.key_metric_name]
+                    if current_val_metric > engine.state.best_metric:
+                        print(f"Got new best metric of {engine.state.key_metric_name}: {current_val_metric}")
+                        engine.state.best_metric = current_val_metric
+                        engine.state.best_metric_epoch = engine.state.epoch
+
+        if handlers is not None and len(handlers) > 0:
+            assert isinstance(handlers, (list, tuple)), "handlers must be a chain."
```

Comment:
If `handlers` is not iterable the for loop below will raise an exception so this check isn't strictly necessary and doesn't permit other types besides list or tuple.

Reply:
removed the assert in the latest patch.
Thanks.